Sage-Bionetworks/schematic
# Lightly refactor manifest generation API endpoint

## Changelog
### To make API routes production ready: 
* Change route name of generating a manifest from `/manifest/generate` to `/manifest`
* Remove access_token parameter and use header to pass token. If user does not provide a token, the API would return 401 error. Related to: https://sagebionetworks.jira.com/browse/DCA-117

### To make code more maintainable: 
* Decoupled long function of manifest generation in `routes.py`
* Added unit tests for manifest generation. `test_routes.py` will be a place for handling unit tests for functions in routes.py
* Simplified the functions of testing generating an existing manifest in `test_api.py`. The old tests have a lot of if else statements that are difficult to read. 
* Refomatted `import` statements using `isort`
* Used black to reformat new tests in `test_routes.py`

## How to (manually) test the PR
Launch API using `poetry run python3 run_api.py`. Here's a list of tests that I have done: 
* Generating a new manifest as google sheet or excel format by using various data type: 
     * schema_url: https://raw.githubusercontent.com/Sage-Bionetworks/schematic/develop/tests/data/example.model.jsonld
     * title: New Example
     * data_type: Patient (OR Biospecimen)
     * output_format: excel (Or google; Or empty)
* Generating an existing manifest in google sheet, excel, and dataframe format. If generating a manifest as an excel format or google sheet format, you should be able to see the title of manifest as `New Example.Patient.manifest.xlsx`: 
     * schema_url: https://raw.githubusercontent.com/Sage-Bionetworks/schematic/develop/tests/data/example.model.jsonld
     * title: New Example
     * data_type: Patient
     * dataset_id: syn51537635
     * asset_view: syn51078429
     * output_format: excel (Or google; Or dataframe) 
* Generating all manifests by using `all manifests` parameter and return three google sheets with three different components 
     * schema_url: https://raw.githubusercontent.com/Sage-Bionetworks/schematic/develop/tests/data/example.model.jsonld
     * title: New Example
     * data_type: all manifests
     * output_format: google
 * Generating all manifests by using `all manifests` parameter and return one Patient excel spreadsheet
      * schema_url: https://raw.githubusercontent.com/Sage-Bionetworks/schematic/develop/tests/data/example.model.jsonld
     * title: New Example
     * data_type: all manifests
     * output_format: excel
* Generating multiple existing manifests as google sheets or excels. If generating as an excel spreadsheet, only one file gets returned. 
     * schema_url: https://raw.githubusercontent.com/Sage-Bionetworks/schematic/develop/tests/data/example.model.jsonld
     * title: New Example
     * data_type: Patient, Biospecimen
     * dataset_id: syn51547836, syn51547837
     * asset_view: syn51547844
     * output_format: google (Or excel)
